x/gamm/twap/api_test.go

```diff
@@ -0,0 +1,204 @@
+
+	blankRecord := types.TwapRecord{}
+	defaultTime := s.Ctx.BlockTime()
+
+	tPlusOneSec := defaultTime.Add(time.Second)
+
+	tests := map[string]struct {
+		// if start record is blank, don't do any sets
+		startRecord types.TwapRecord
+		// We set it to have the updated time
+		expRecord  types.TwapRecord
+		time       time.Time
+		poolId     uint64
+		quoteDenom string
+		baseDenom  string
+		expError   bool
+	}{
+		"no record (wrong pool ID)": {blankRecord, blankRecord, defaultTime, 4, denomA, denomB, true},
+		"default record":            {blankRecord, initStartRecord, defaultTime, 1, denomA, denomB, false},
+		"one second later record":   {blankRecord, recordWithUpdatedAccum(initStartRecord, OneSec, OneSec), tPlusOneSec, 1, denomA, denomB, false},
+		"idempotent overwrite":      {initStartRecord, initStartRecord, defaultTime, 1, denomA, denomB, false},
+		"idempotent overwrite2":     {initStartRecord, recordWithUpdatedAccum(initStartRecord, OneSec, OneSec), tPlusOneSec, 1, denomA, denomB, false},
+		"diff spot price": {zeroAccumTenPoint1Record,
+			recordWithUpdatedAccum(zeroAccumTenPoint1Record, OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+			tPlusOneSec, 1, denomA, denomB, false},
+		// TODO: Overflow
+	}
+	for name, tc := range tests {
+		s.Run(name, func() {
+			// setup time
+			s.Ctx = s.Ctx.WithBlockTime(tc.time)
+			tc.expRecord.Time = tc.time
+
+			// setup record
+			initSetRecord := tc.startRecord
+			if (tc.startRecord == types.TwapRecord{}) {
+				initSetRecord = initStartRecord
+			}
+			s.twapkeeper.StoreNewRecord(s.Ctx, initSetRecord)
+
+			actualRecord, err := s.twapkeeper.GetBeginBlockAccumulatorRecord(s.Ctx, tc.poolId, tc.baseDenom, tc.quoteDenom)
+
+			if tc.expError {
+				s.Require().Error(err)
+				return
+			}
+			s.Require().NoError(err)
+			s.Require().Equal(tc.expRecord, actualRecord)
+		})
+	}
+}
+
+func (s *TestSuite) TestGetArithmeticTwap() {
+	type getTwapInput struct {
+		poolId          uint64
+		quoteAssetDenom string
+		baseAssetDenom  string
+		startTime       time.Time
+		endTime         time.Time
+	}
+
+	makeSimpleTwapInput := func(startTime time.Time, endTime time.Time, isQuoteTokenA bool) getTwapInput {
+		quoteAssetDenom, baseAssetDenom := denom0, denom1
+		if isQuoteTokenA {
+			baseAssetDenom, quoteAssetDenom = quoteAssetDenom, baseAssetDenom
+		}
+		return getTwapInput{1, quoteAssetDenom, baseAssetDenom, startTime, endTime}
+	}
+
+	quoteAssetA := true
+	quoteAssetB := false
+	tPlusOne := baseTime.Add(time.Second)
+	// base record is a record with t=baseTime, sp0=10, sp1=.1, accumulators set to 0
+	baseRecord := newTwapRecordWithDefaults(baseTime, sdk.NewDec(10), sdk.ZeroDec(), sdk.ZeroDec())
+	// record with t=baseTime+10, sp0=5, sp1=.2, accumulators updated from baseRecord
+	tPlus10sp5Record := newTwapRecordWithDefaults(
+		baseTime.Add(10*time.Second), sdk.NewDec(5), OneSec.MulInt64(10*10), OneSec)
```

Comment:
I know `OneSec` was probably chosen with the goal of having any value here but I think, in terms of abstraction, it might be better to avoid using only `OnseSec` because the accumulator also includes the price

Reply:
Wait no, this is `accumulator = (one second as number) * (10 seconds) * (10 spot price)`

Reply:
This is true for `accum0` but there is also `accum1` which only gets gets `OneSec` I think

Reply:
Ah yeah, I should probably add a comment here: that one is `accumulator = (one second as number) * (10 seconds) * (.1 spot price)`

These are explicitly chosen to be correct, but incredibly not obvious. Let me define these in their own line w/ a comment